src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java

```diff
@@ -39,90 +39,216 @@
+                                            try {
+                                                return new ArchivedRecordingInfo(
+                                                        subdirectory,
+                                                        file,
+                                                        webServer.getArchivedReportURL(file),
+                                                        webServer.getArchivedDownloadURL(file));
+                                            } catch (SocketException
+                                                    | UnknownHostException
+                                                    | URISyntaxException e) {
+                                                logger.warn(e);
+                                                return null;
+                                            }
+                                        })
+                                .filter(Objects::nonNull)
+                                .collect(Collectors.toList());
+                archivedRecordings.addAll(temp);
+            }
+            future.complete(archivedRecordings);
+        } catch (ArchivePathException e) {
+            future.completeExceptionally(e);
+        }
 
-        return saveName;
+        return future;
     }
 
-    public void deleteRecording(ConnectionDescriptor connectionDescriptor, String recordingName)
-            throws Exception {
+    public Future<Path> getRecordingPath(String recordingName) throws Exception {
+        CompletableFuture<Path> future = new CompletableFuture<>();
 
-        targetConnectionManager.executeConnectedTask(
-                connectionDescriptor,
-                connection -> {
-                    Optional<IRecordingDescriptor> descriptor =
-                            this.getDescriptorByName(connection, recordingName);
+        try {
+            Path archivedRecording = null;
+            Boolean recordingFound = false;
+            List<String> subdirectories = this.fs.listDirectoryChildren(recordingsPath);
+            for (String subdirectory : subdirectories) {
+                List<String> files =
+                        this.fs.listDirectoryChildren(recordingsPath.resolve(subdirectory));
 
-                    if (descriptor.isPresent()) {
-                        connection.getService().close(descriptor.get());
-                        reportService.delete(connectionDescriptor, recordingName);
-                    } else {
-                        throw new RecordingNotFoundException(recordingName);
+                for (String file : files) {
+                    if (recordingName.equals(file)) {
+                        archivedRecording = recordingsPath.resolve(subdirectory + "/" + file);
+                        recordingFound = true;
+                        break;
                     }
-                    return null;
-                });
+                }
+
+                if (recordingFound) {
```

Comment:
It's easily abused so usually best avoided, but in this case a labelled break:

(see https://docs.oracle.com/javase/tutorial/java/nutsandbolts/branch.html)

could be used instead of this `recordingFound` flag to break out of the outer loop from the inner loop.

Reply:
Alternatively you could extract this piece of logic out into a small helper method, and use `return` to break the control flow out from the inner loop back up to the callsite.